core/blockchain.go

```diff
@@ -964,8 +966,25 @@ func (bc *BlockChain) WriteBlockWithState(block *types.Block, receipts []*types.
 	reorg := externTd.Cmp(localTd) > 0
 	currentBlock = bc.CurrentBlock()
 	if !reorg && externTd.Cmp(localTd) == 0 {
-		// Split same-difficulty blocks by number, then at random
-		reorg = block.NumberU64() < currentBlock.NumberU64() || (block.NumberU64() == currentBlock.NumberU64() && mrand.Float64() < 0.5)
+		// Split same-difficulty blocks by number, then preferentially select
+		// the block generated by the local miner as the canonical block.
+		if block.NumberU64() < currentBlock.NumberU64() {
+			reorg = true
+		} else if block.NumberU64() == currentBlock.NumberU64() {
+			currentAuthor, err := bc.engine.Author(currentBlock.Header())
+			if err != nil {
+				return NonStatTy, err
+			}
+			blockAuthor, err := bc.engine.Author(block.Header())
+			if err != nil {
+				return NonStatTy, err
+			}
+			var currentLocal, newLocal bool
```

Comment:
Nitpick: if you added `currentAuthor` and `blockAuthor`, then please also use `currentLocal` and `blockLocal` (instead of `newLocal`).